File: 2016/tools/xlsxread/src/writer.rs
/// SQLite writer: DDL setup, batched INSERT OR REPLACE, VACUUM, stats output.
///
/// Mirrors build-lib.js createDb + the transaction loop in each build-database*.js.
/// Stats output lines match the JS stdout exactly so existing CI log-greps still work.
///
/// Every dataset writes the same canonical table (see `crate::schema`), so there
/// is exactly one insert path. Columns a dataset carries no data for bind NULL.
use std::fs;
use std::path::Path;

use rusqlite::{params_from_iter, Connection, ToSql};

use crate::error::BuildError;
use crate::schema;
use crate::transform::ParsedRow;

// ---------------------------------------------------------------------------
// DB initialisation — mirrors build-lib.js createDb (delete + recreate)
// ---------------------------------------------------------------------------

/// Open (or recreate) the output SQLite database, execute the canonical DDL,
/// and return the open connection ready for inserts.
pub fn open_db(db_path: &Path) -> Result<Connection, BuildError> {
    // Mirror Node behaviour: delete existing file before creating (build-lib.js:54)
    if db_path.exists() {
        fs::remove_file(db_path).map_err(|e| BuildError::Io {
            path: db_path.display().to_string(),
            source: e,
        })?;
    }

    // Ensure parent directory exists
    if let Some(parent) = db_path.parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent).map_err(|e| BuildError::Io {
                path: parent.display().to_string(),
                source: e,
            })?;
        }
    }

    let conn = Connection::open(db_path)?;
    conn.execute_batch(schema::DDL)?;
    Ok(conn)
}

// ---------------------------------------------------------------------------
// Insert a single parsed row
// ---------------------------------------------------------------------------

/// Bind every field from `row` into the canonical INSERT and execute it.
///
/// Parameter order is `schema::IDENTITY_FIELDS` followed by
/// `schema::SCORE_FIELDS`. Subjects absent from `row.scores` — and the two
/// identity columns only the thptqg2016 layouts populate — bind NULL.
pub fn insert_row(conn: &Connection, row: &ParsedRow) -> Result<(), BuildError> {
    let mut params: Vec<Box<dyn ToSql>> = Vec::with_capacity(schema::PARAM_COUNT);

    params.push(Box::new(row.so_bao_danh.clone()));
    params.push(Box::new(row.ho_ten.clone()));
    params.push(Box::new(row.ho_ten_ascii.clone()));
    params.push(Box::new(row.ngay_sinh.clone()));
    params.push(Box::new(row.ten_cum_thi.clone()));
    params.push(Box::new(row.gioi_tinh.clone()));

    for field in schema::SCORE_FIELDS {
        let val: Option<f64> = row.scores.get(*field).copied();
        params.push(Box::new(val));
    }

    debug_assert_eq!(params.len(), schema::PARAM_COUNT);

    conn.execute(
        schema::INSERT_SQL,
        params_from_iter(params.iter().map(|p| p.as_ref())),
    )?;
    Ok(())
}

// ---------------------------------------------------------------------------
// Post-build: VACUUM + stats output
// ---------------------------------------------------------------------------

/// Run VACUUM and print statistics lines that mirror the Node scripts' stdout.
/// The exact prefix tokens ("Source data rows", "DB rows", "Size:") are preserved
/// so any log-grep in the deploy pipeline keeps working.
#[allow(clippy::too_many_arguments)]
pub fn finish_db(
    conn: &Connection,
    db_path: &Path,
    source_rows: u64,
    skipped: u64,
    errors: u64,
    dataset_label: &str,
    _file_count: usize,
    is_old2: bool,
) -> Result<(), BuildError> {
    conn.execute_batch("VACUUM")?;

    let db_count: i64 = conn.query_row("SELECT COUNT(*) FROM student", [], |row| row.get(0))?;

    let insertable = source_rows - skipped;

    println!();
    if is_old2 {
        println!("Source non-blank data rows:      {source_rows}");
        println!("  skipped (empty/non-numeric SBD): {skipped}");
    } else {
        println!("Source data rows (post-header):  {source_rows}");
        if dataset_label.contains("old") {
            println!("  skipped (empty/non-numeric SBD): {skipped}");
        } else {
            println!("  skipped (empty/invalid):        {skipped}");
        }
    }
    println!("  insertable:                     {insertable}");
    println!("  insert errors:                  {errors}");
    println!("DB rows (distinct SBD):           {db_count}");

    if !dataset_label.contains("old") && errors == 0 {
        let gap = insertable as i64 - db_count;
        if gap == 0 {
            println!("Audit: OK — every source row made it in.");
        } else {
            println!("Audit: {gap} row(s) collapsed (duplicate SBDs overwriting).");
        }
    }

    let sz = fs::metadata(db_path).map(|m| m.len()).unwrap_or(0);
    println!("Size: {:.1} MB", sz as f64 / 1024.0 / 1024.0);

    Ok(())
}

File: parser/internal/writer/writer.go
// Package writer handles SQLite output: DDL setup, INSERT OR REPLACE, VACUUM
// and the stats block.
//
// Every dataset writes the same canonical table (internal/schema), so there is
// exactly one insert path. Columns a dataset carries no data for bind NULL.
//
// The stats lines are the operator-facing output of the build, and
// docs/deployment-guide.md points at the per-file row counts for
// troubleshooting — do not reword them casually.
package writer

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/tiennm99/thptqg/parser/internal/schema"
	"github.com/tiennm99/thptqg/parser/internal/sqlitedb"
	"github.com/tiennm99/thptqg/parser/internal/transform"
)

// OpenDB deletes any existing database at dbPath, recreates it, and executes the
// canonical DDL.
//
// Deleting the file rather than issuing DROP TABLE has a consequence worth
// knowing: a concurrent reader sees the file vanish mid-rebuild rather than a
// transactional swap.
func OpenDB(dbPath string) (*sql.DB, error) {
	if _, err := os.Stat(dbPath); err == nil {
		if err := os.Remove(dbPath); err != nil {
			return nil, fmt.Errorf("remove existing db %s: %w", dbPath, err)
		}
	}
	if parent := filepath.Dir(dbPath); parent != "" && parent != "." {
		if err := os.MkdirAll(parent, 0o755); err != nil {
			return nil, fmt.Errorf("create %s: %w", parent, err)
		}
	}

	db, err := sql.Open(sqlitedb.DriverName, dbPath)
	if err != nil {
		return nil, fmt.Errorf("open db %s: %w", dbPath, err)
	}
	if _, err := db.Exec(schema.DDL); err != nil {
		db.Close()
		return nil, fmt.Errorf("execute DDL: %w", err)
	}
	return db, nil
}

// Inserter wraps the canonical INSERT, prepared once per transaction rather
// than per row.
type Inserter struct{ stmt *sql.Stmt }

// Prepare compiles the canonical INSERT against tx.
func Prepare(tx *sql.Tx) (*Inserter, error) {
	stmt, err := tx.Prepare(schema.InsertSQL)
	if err != nil {
		return nil, fmt.Errorf("prepare insert: %w", err)
	}
	return &Inserter{stmt: stmt}, nil
}

// Close releases the prepared statement.
func (i *Inserter) Close() error { return i.stmt.Close() }

// Insert binds one parsed row and executes the INSERT.
//
// Parameter order is IdentityFields then ScoreFields. Subjects absent from
// row.Scores — and the two identity columns only the 2016 layouts populate —
// bind NULL.
func (i *Inserter) Insert(row *transform.ParsedRow) error {
	args := make([]any, 0, schema.ParamCount)
	args = append(args,
		row.SoBaoDanh,
		row.HoTen,
		row.HoTenAscii,
		nullableString(row.NgaySinh),
		nullableString(row.TenCumThi),
		nullableString(row.GioiTinh),
	)
	for _, field := range schema.ScoreFields {
		if v, ok := row.Scores[field]; ok {
			args = append(args, v)
		} else {
			args = append(args, nil)
		}
	}
	if len(args) != schema.ParamCount {
		return fmt.Errorf("built %d params, want %d", len(args), schema.ParamCount)
	}
	if _, err := i.stmt.Exec(args...); err != nil {
		return err
	}
	return nil
}

func nullableString(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

// Stats carries the counters the build loop accumulates.
type Stats struct {
	SourceRows uint64
	Skipped    uint64
	Errors     uint64
}

// BuildNameIndex fills name_word from the student rows, one entry per distinct
// word of each ASCII name.
//
// A second pass rather than a write alongside each insert: a repeated exam
// number replaces its earlier row, and the words of the row it replaced would
// otherwise stay behind pointing at a name that is no longer there.
func BuildNameIndex(db *sql.DB) error {
	rows, err := db.Query("SELECT so_bao_danh, ho_ten_ascii FROM student")
	if err != nil {
		return fmt.Errorf("read names: %w", err)
	}
	defer rows.Close()

	tx, err := db.Begin()
	if err != nil {
		return fmt.Errorf("begin name index: %w", err)
	}
	stmt, err := tx.Prepare(schema.NameWordInsertSQL)
	if err != nil {
		tx.Rollback()
		return fmt.Errorf("prepare name index: %w", err)
	}

	var words uint64
	seen := make(map[string]struct{}, 8)
	for rows.Next() {
		var sbd, ascii string
		if err := rows.Scan(&sbd, &ascii); err != nil {
			tx.Rollback()
			return fmt.Errorf("scan name: %w", err)
		}
		clear(seen)
		for _, w := range strings.Fields(ascii) {
			if _, dup := seen[w]; dup {
				continue
			}
			seen[w] = struct{}{}
			if _, err := stmt.Exec(w, sbd, ascii); err != nil {
				tx.Rollback()
				return fmt.Errorf("insert name word: %w", err)
			}
			words++
		}
	}
	if err := rows.Err(); err != nil {
		tx.Rollback()
		return fmt.Errorf("read names: %w", err)
	}
	if err := stmt.Close(); err != nil {
		tx.Rollback()
		return err
	}
	if err := tx.Commit(); err != nil {
		return fmt.Errorf("commit name index: %w", err)
	}

	if _, err := db.Exec(schema.PostLoadSQL); err != nil {
		return fmt.Errorf("post-load statements: %w", err)
	}
	fmt.Printf("Name index:                       %d words\n", words)
	return nil
}

// Finish builds the derived tables, runs VACUUM and prints the stats block.
//
// VACUUM must run AFTER the transaction commits — SQLite refuses it inside one
// — and after the name index, so the file is laid out in one pass.
func Finish(db *sql.DB, dbPath string, st Stats) error {
	if err := BuildNameIndex(db); err != nil {
		return err
	}
	if _, err := db.Exec("VACUUM"); err != nil {
		return fmt.Errorf("vacuum: %w", err)
	}

	var dbCount int64
	if err := db.QueryRow("SELECT COUNT(*) FROM student").Scan(&dbCount); err != nil {
		return fmt.Errorf("count rows: %w", err)
	}

	insertable := st.SourceRows - st.Skipped

	fmt.Println()
	fmt.Printf("Source data rows (post-header):  %d\n", st.SourceRows)
	fmt.Printf("  skipped (empty/invalid):        %d\n", st.Skipped)
	fmt.Printf("  insertable:                     %d\n", insertable)
	fmt.Printf("  insert errors:                  %d\n", st.Errors)
	fmt.Printf("DB rows (distinct SBD):           %d\n", dbCount)

	if st.Errors == 0 {
		gap := int64(insertable) - dbCount
		if gap == 0 {
			fmt.Println("Audit: OK — every source row made it in.")
		} else {
			fmt.Printf("Audit: %d row(s) collapsed (duplicate SBDs overwriting).\n", gap)
		}
	}

	var size int64
	if fi, err := os.Stat(dbPath); err == nil {
		size = fi.Size()
	}
	fmt.Printf("Size: %.1f MB\n", float64(size)/1024.0/1024.0)
	return nil
}
